satpy/resample.py

```diff
@@ -823,11 +823,187 @@ def compute(self, data, expand=True, **kwargs):
                             coords=coords or None)
 
 
+class BucketResamplerBase(BaseResampler):
+    """Base class for bucket resampling which implements averaging.
+    """
+
+    def __init__(self, source_geo_def, target_geo_def):
+        super(BucketResamplerBase, self).__init__(source_geo_def, target_geo_def)
+        self.resampler = None
+
+    def precompute(self, **kwargs):
+        """Create X and Y indices and store them for later use."""
+        LOG.debug("Initializing bucket resampler.")
+        source_lons, source_lats = self.source_geo_def.get_lonlats(
+            chunks=CHUNK_SIZE)
+        self.resampler = bucket.BucketResampler(self.target_geo_def,
+                                                source_lons,
+                                                source_lats)
+
+    def compute(self, data, **kwargs):
+        """Call the resampling."""
+        raise NotImplementedError("Use the sub-classes")
+
+    def resample(self, data, **kwargs):
+        """Resample `data` by calling `precompute` and `compute` methods.
+
+        Args:
+            data (xarray.DataArray): Data to be resampled
+
+        Returns (xarray.DataArray): Data resampled to the target area
+        """
+        self.precompute(**kwargs)
+        attrs = data.attrs.copy()
+        data_arr = data.data
+        if data.ndim == 3 and data.dims[0] == 'bands':
+            dims = ('bands', 'y', 'x')
+        # Both one and two dimensional input data results in 2D output
+        elif data.ndim in (1, 2):
+            dims = ('y', 'x')
+        else:
+            dims = data.dims
```

Comment:
Is this dimension handling correct? We can assume that they have at least a `y` dimension in satpy. Do you preserve the dimension order in the pyresample resampler? This isn't a big deal, but would appreciate a second glance at simplifying this logic.

Reply:
In Pyresample `BucketResampler` works with 2D input data that is raveled to 1D before resampling. The output is one 2D Dask array (`bucket_fraction` returns a dict of 2D Dask arrays) with `'y'` as first dimension and the shape of the target area. 3D input is handled on Satpy's side.